# FE Homework 5
**FBA Quant 이의형**

---

## Problem 1.

Reproduce the payer-swaption lattices in the Coursera Lecture.  

Now experiment with different values of b (but re-calibrating each time so that the spot rate curve remains unchanged) to see how sensitive the swaption price is to the particular value of b that you choose. 

(Among other things, this should highlight the importance of calibrating models correctly and understanding whether or not your model is appropriate for the problem at hand. For example, we know that option prices are sensitive to volatility and since b is a volatility parameter in the BDT model, it is clearly important to calibrate it accurately. In fact we may wish to choose and calibrate a separate bi for each time period. If we were pricing a swaption where the underlying swap expired after 10 years instead of just 10 months we would also want to consider whether or not a model that cannot, for example, incorporate mean-reversion should be used for such a task.)


### 1-1. Find a

In [1]:
import pandas as pd
import numpy as np

mkt_spot = [7.30, 7.90, 8.98, 9.37, 12.01, 11.57, 12.66, 12.35, 12.66, 14.86, 14.18, 15.21, 14.71, 13.02] # 우리는 7.3밖에 모름
a = [ 5 for i in range(14)]
b = 0.005

In [2]:
short_lattice = pd.DataFrame(columns = [ i for i in range(14)], index = [ i for i in range(13, -1, -1) ])
short_lattice.loc[0,0] = a[0]
for t in range(1, len(short_lattice.columns)):
    for s in range(t+1):
        short_lattice.loc[s,t] = a[t]*np.exp(b*s)
short_lattice

0         1         2         3         4         5         6         7   \
13  NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
12  NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
11  NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
10  NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
9   NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
8   NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
7   NaN       NaN       NaN       NaN       NaN       NaN       NaN  5.178099   
6   NaN       NaN       NaN       NaN       NaN       NaN  5.152273  5.152273   
5   NaN       NaN       NaN       NaN       NaN  5.126576  5.126576  5.126576   
4   NaN       NaN       NaN       NaN  5.101007  5.101007  5.101007  5.101007   
3   NaN       NaN       NaN  5.075565  5.075565  5.075565  5.075565  5.075565   
2   NaN       NaN  5.050251  5.050251  5.050251  5.050251  5.050251  5.050251   
1   NaN  5.025063  5.025063  5.025063  5.025063  5.025063  5.025063  5.025063   
0     5       5.0       5.0       5.0       5.0       5.0       5.0       5.0   

          8         9         10        11        12        13  
13       NaN       NaN       NaN       NaN       NaN  5.335795  
12       NaN       NaN       NaN       NaN  5.309183  5.309183  
11       NaN       NaN       NaN  5.282703  5.282703  5.282703  
10       NaN       NaN  5.256355  5.256355  5.256355  5.256355  
9        NaN  5.230139  5.230139  5.230139  5.230139  5.230139  
8   5.204054  5.204054  5.204054  5.204054  5.204054  5.204054  
7   5.178099  5.178099  5.178099  5.178099  5.178099  5.178099  
6   5.152273  5.152273  5.152273  5.152273  5.152273  5.152273  
5   5.126576  5.126576  5.126576  5.126576  5.126576  5.126576  
4   5.101007  5.101007  5.101007  5.101007  5.101007  5.101007  
3   5.075565  5.075565  5.075565  5.075565  5.075565  5.075565  
2   5.050251  5.050251  5.050251  5.050251  5.050251  5.050251  
1   5.025063  5.025063  5.025063  5.025063  5.025063  5.025063  
0        5.0       5.0       5.0       5.0       5.0       5.0

In [3]:
el_price = pd.DataFrame(columns = [ i for i in range(15)], index = [ i for i in range(14, -1, -1) ])
el_price.loc[0,0] = 1
for t in range(1, len(el_price.columns)):
    for s in range(t+1):
        if s == t:
            el_price.loc[s,t] = 0.5 * el_price.loc[s-1,t-1] / (1 + short_lattice.loc[s-1,t-1]/100)
        elif s == 0:
            el_price.loc[s,t] = 0.5 * el_price.loc[s,t-1] / (1 + short_lattice.loc[s,t-1]/100)
        else:
            el_price.loc[s,t] = 0.5*((el_price.loc[s-1,t-1]/(1+short_lattice.loc[s-1,t-1]/100))+(el_price.loc[s,t-1]/(1+short_lattice.loc[s,t-1]/100)))
el_price

0        1         2         3         4         5         6         7   \
14  NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
13  NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
12  NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
11  NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
10  NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
9   NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
8   NaN      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
7   NaN      NaN       NaN       NaN       NaN       NaN       NaN  0.005524   
6   NaN      NaN       NaN       NaN       NaN       NaN  0.011618  0.038698   
5   NaN      NaN       NaN       NaN       NaN  0.024427  0.069748  0.116177   
4   NaN      NaN       NaN       NaN  0.051345  0.122192  0.174476  0.193769   
3   NaN      NaN       NaN  0.107902  0.205454  0.244501  0.232774  0.193909   
2   NaN      NaN  0.226703  0.323784  0.308292  0.244618  0.174685  0.116429   
1   NaN  0.47619  0.453461  0.323862  0.205602  0.122368  0.069916  0.038838   
0     1  0.47619  0.226757   0.10798  0.051419  0.024485   0.01166  0.005552   

          8         9         10        11        12        13        14  
14       NaN       NaN       NaN       NaN       NaN       NaN   0.00003  
13       NaN       NaN       NaN       NaN       NaN  0.000064  0.000423  
12       NaN       NaN       NaN       NaN  0.000134  0.000827  0.002753  
11       NaN       NaN       NaN  0.000282  0.001608  0.004969  0.011028  
10       NaN       NaN  0.000593  0.003103  0.008854  0.018248  0.030376  
9        NaN  0.001248  0.005937  0.015532  0.029553  0.045687   0.06085  
8   0.002626  0.011244  0.026746  0.046653  0.066584  0.082357   0.09142  
7   0.021027   0.04502    0.0714  0.093421  0.106677   0.10997  0.104645  
6   0.073657  0.105148  0.125086  0.130948  0.124622  0.110131  0.091709  
5   0.147438  0.157875  0.150267  0.131106  0.106961  0.082718  0.061236  
4   0.184453  0.158027  0.125358  0.093761   0.06694  0.046021  0.030666  
3   0.147687  0.105453  0.071711  0.046937   0.02979  0.018435  0.011169  
2   0.073906  0.045238  0.026921  0.015664  0.008949  0.005035  0.002797  
1   0.021134   0.01132  0.005989  0.003137  0.001629   0.00084  0.000431  
0   0.002644  0.001259    0.0006  0.000285  0.000136  0.000065  0.000031

In [4]:
bdt_zcb = el_price.sum(axis=0)
bdt_spot = 100 * (((1/bdt_zcb)**(1/bdt_zcb.index))-1)
bdt_spot

0          0.0
1          5.0
2     5.006265
3     5.012539
4     5.018824
5     5.025118
6     5.031422
7     5.037736
8      5.04406
9     5.050394
10    5.056738
11    5.063092
12    5.069455
13    5.075829
14    5.082213
dtype: object

In [5]:
difference = ((bdt_spot.iloc[1:] - mkt_spot)**2).sum()
difference

747.1388585215972

In [6]:
from scipy.optimize import minimize

mkt_spot = [7.3, 7.62, 8.1, 8.45, 9.2, 9.64, 10.12, 10.45, 10.75, 11.22, 11.55, 11.92, 12.2, 12.32]
b = 0.005

def obj_function(a):
    short_lattice = pd.DataFrame(columns = [ i for i in range(14)], index = [ i for i in range(13, -1, -1) ])
    short_lattice.loc[0,0] = a[0]
    for t in range(1, len(short_lattice.columns)):
        for s in range(t+1):
            short_lattice.loc[s,t] = a[t]*np.exp(b*s)

    el_price = pd.DataFrame(columns = [ i for i in range(15)], index = [ i for i in range(14, -1, -1) ])
    el_price.loc[0,0] = 1
    for t in range(1, len(el_price.columns)):
        for s in range(t+1):
            if s == t:
                el_price.loc[s,t] = 0.5 * el_price.loc[s-1,t-1] / (1 + short_lattice.loc[s-1,t-1]/100)
            elif s == 0:
                el_price.loc[s,t] = 0.5 * el_price.loc[s,t-1] / (1 + short_lattice.loc[s,t-1]/100)
            else:
                el_price.loc[s,t] = 0.5*((el_price.loc[s-1,t-1]/(1+short_lattice.loc[s-1,t-1]/100))+(el_price.loc[s,t-1]/(1+short_lattice.loc[s,t-1]/100)))

    bdt_zcb = el_price.sum(axis=0)
    bdt_spot = 100 * (((1/bdt_zcb)**(1/bdt_zcb.index))-1)

    return ((bdt_spot.iloc[1:] - mkt_spot)**2).sum()

a_init = [ 5 for i in range(14)]
b_ = [(0,20) for i in range(len(a_init))]

result = minimize(obj_function, a_init, bounds = b_)

In [7]:
print(f'제곱오차: {result.fun}')
a = result.x
a

제곱오차: 1.5753314519321333e-08


array([ 7.30000426,  7.92110169,  9.0211975 ,  9.43561236, 12.13037944,
       11.71922742, 12.85003131, 12.56572679, 12.91925815, 15.19450552,
       14.53762509, 15.63516143, 15.15332319, 13.44884627])

### 2-2. Swaption Pricing

In [8]:
# 1백만 달러 가정
b = 0.005
fix = 11.65

short_lattice = pd.DataFrame(columns = [ i for i in range(14)], index = [ i for i in range(13, -1, -1) ])
short_lattice.loc[0,0] = a[0]
for t in range(1, len(short_lattice.columns)):
    for s in range(t+1):
        short_lattice.loc[s,t] = a[t]*np.exp(b*s)
short_lattice

0         1         2         3          4          5          6   \
13       NaN       NaN       NaN       NaN        NaN        NaN        NaN   
12       NaN       NaN       NaN       NaN        NaN        NaN        NaN   
11       NaN       NaN       NaN       NaN        NaN        NaN        NaN   
10       NaN       NaN       NaN       NaN        NaN        NaN        NaN   
9        NaN       NaN       NaN       NaN        NaN        NaN        NaN   
8        NaN       NaN       NaN       NaN        NaN        NaN        NaN   
7        NaN       NaN       NaN       NaN        NaN        NaN        NaN   
6        NaN       NaN       NaN       NaN        NaN        NaN  13.241373   
5        NaN       NaN       NaN       NaN        NaN  12.015901  13.175331   
4        NaN       NaN       NaN       NaN  12.375429  11.955972  13.109619   
3        NaN       NaN       NaN  9.578213  12.313707  11.896341  13.044235   
2        NaN       NaN  9.111862  9.530442  12.252292  11.837008  12.979176   
1        NaN  7.960806  9.066416  9.482909  12.191183   11.77797  12.914442   
0   7.300004  7.921102  9.021197  9.435612  12.130379  11.719227  12.850031   

           7          8          9          10         11         12  \
13        NaN        NaN        NaN        NaN        NaN        NaN   
12        NaN        NaN        NaN        NaN        NaN  16.090352   
11        NaN        NaN        NaN        NaN  16.519183  16.010101   
10        NaN        NaN        NaN  15.282985  16.436793  15.930251   
9         NaN        NaN  15.893876  15.206761  16.354814  15.850798   
8         NaN  13.446503  15.814605  15.130917  16.273244  15.771742   
7   13.013314  13.379438  15.735729  15.055451  16.192081   15.69308   
6    12.94841  13.312708  15.657247  14.980362  16.111323  15.614811   
5    12.88383  13.246311  15.579156  14.905647  16.030967  15.536931   
4   12.819571  13.180244  15.501455  14.831305  15.951013  15.459441   
3   12.755633  13.114508  15.424141  14.757333  15.871457  15.382336   
2   12.692014  13.049099  15.347213  14.683731  15.792297  15.305617   
1   12.628713  12.984016  15.270668  14.610495  15.713533   15.22928   
0   12.565727  12.919258  15.194506  14.537625  15.635161  15.153323   

           13  
13  14.352058  
12  14.280476  
11  14.209252  
10  14.138383  
9   14.067868  
8   13.997704  
7    13.92789  
6   13.858425  
5   13.789305  
4   13.720531  
3     13.6521  
2   13.584009  
1   13.516259  
0   13.448846

In [9]:
fix = fix/100
short_lattice = short_lattice/100
short_lattice

0         1         2         3         4         5         6   \
13    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
12    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
11    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
10    NaN       NaN       NaN       NaN       NaN       NaN       NaN   
9     NaN       NaN       NaN       NaN       NaN       NaN       NaN   
8     NaN       NaN       NaN       NaN       NaN       NaN       NaN   
7     NaN       NaN       NaN       NaN       NaN       NaN       NaN   
6     NaN       NaN       NaN       NaN       NaN       NaN  0.132414   
5     NaN       NaN       NaN       NaN       NaN  0.120159  0.131753   
4     NaN       NaN       NaN       NaN  0.123754   0.11956  0.131096   
3     NaN       NaN       NaN  0.095782  0.123137  0.118963  0.130442   
2     NaN       NaN  0.091119  0.095304  0.122523   0.11837  0.129792   
1     NaN  0.079608  0.090664  0.094829  0.121912   0.11778  0.129144   
0   0.073  0.079211  0.090212  0.094356  0.121304  0.117192    0.1285   

          7         8         9         10        11        12        13  
13       NaN       NaN       NaN       NaN       NaN       NaN  0.143521  
12       NaN       NaN       NaN       NaN       NaN  0.160904  0.142805  
11       NaN       NaN       NaN       NaN  0.165192  0.160101  0.142093  
10       NaN       NaN       NaN   0.15283  0.164368  0.159303  0.141384  
9        NaN       NaN  0.158939  0.152068  0.163548  0.158508  0.140679  
8        NaN  0.134465  0.158146  0.151309  0.162732  0.157717  0.139977  
7   0.130133  0.133794  0.157357  0.150555  0.161921  0.156931  0.139279  
6   0.129484  0.133127  0.156572  0.149804  0.161113  0.156148  0.138584  
5   0.128838  0.132463  0.155792  0.149056   0.16031  0.155369  0.137893  
4   0.128196  0.131802  0.155015  0.148313   0.15951  0.154594  0.137205  
3   0.127556  0.131145  0.154241  0.147573  0.158715  0.153823  0.136521  
2    0.12692  0.130491  0.153472  0.146837  0.157923  0.153056   0.13584  
1   0.126287   0.12984  0.152707  0.146105  0.157135  0.152293  0.135163  
0   0.125657  0.129193  0.151945  0.145376  0.156352  0.151533  0.134488

In [10]:
swap_price = short_lattice.loc[9:,:9].copy()
swap_price.iloc[:,-1] = (swap_price.iloc[:,-1]-fix) / (1+swap_price.iloc[:,-1])
for t in swap_price.columns[:-1][::-1]:
    for s in range(t+1):
        swap_price.loc[s,t] = 1/(1+short_lattice.loc[s,t])*((short_lattice.loc[s,t]-fix) + 0.5*(swap_price.loc[s+1,t+1] + swap_price.loc[s, t+1]))
swap_price

0         1         2         3         4         5         6  \
9      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
8      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
7      NaN       NaN       NaN       NaN       NaN       NaN       NaN   
6      NaN       NaN       NaN       NaN       NaN       NaN  0.060941   
5      NaN       NaN       NaN       NaN       NaN  0.056813  0.059021   
4      NaN       NaN       NaN       NaN  0.056025  0.054595  0.057105   
3      NaN       NaN       NaN  0.031084  0.053533  0.052381  0.055194   
2      NaN       NaN  0.003991  0.028388  0.051045  0.050171  0.053286   
1      NaN -0.031811  0.001105  0.025695   0.04856  0.047965  0.051383   
0 -0.07161 -0.034864 -0.001778  0.023004  0.046078  0.045763  0.049484   

          7         8         9  
9       NaN       NaN  0.036619  
8       NaN  0.047823  0.035959  
7  0.053874   0.04668  0.035302  
6  0.052319  0.045539  0.034648  
5  0.050769  0.044403  0.033995  
4  0.049222   0.04327  0.033346  
3   0.04768   0.04214  0.032698  
2  0.046141  0.041014  0.032053  
1  0.044607  0.039892   0.03141  
0  0.043077  0.038773   0.03077

In [11]:
swaption_price = swap_price.copy()
swaption_price[swaption_price<0] = 0
swaption_price = swaption_price.loc[2:,:2]
for t in swaption_price.columns[:-1][::-1]:
    for s in range(t+1):
        swaption_price.loc[s,t] = (swaption_price.loc[s+1,t+1] + swaption_price.loc[s,t+1])*0.5/(1+short_lattice.loc[s,t])

In [12]:
swaption_price

0         1         2
2       NaN       NaN  0.003991
1       NaN   0.00236  0.001105
0  0.001339  0.000512         0

### 1-3. comparison according to b

In [13]:
# 위 프로세스 함수로 정리

from scipy.optimize import minimize

def find_a(mkt_spot, b):
    def obj_function(a):
        short_lattice = pd.DataFrame(columns = [ i for i in range(14)], index = [ i for i in range(13, -1, -1) ])
        short_lattice.loc[0,0] = a[0]
        for t in range(1, len(short_lattice.columns)):
            for s in range(t+1):
                short_lattice.loc[s,t] = a[t]*np.exp(b*s)

        el_price = pd.DataFrame(columns = [ i for i in range(15)], index = [ i for i in range(14, -1, -1) ])
        el_price.loc[0,0] = 1
        for t in range(1, len(el_price.columns)):
            for s in range(t+1):
                if s == t:
                    el_price.loc[s,t] = 0.5 * el_price.loc[s-1,t-1] / (1 + short_lattice.loc[s-1,t-1]/100)
                elif s == 0:
                    el_price.loc[s,t] = 0.5 * el_price.loc[s,t-1] / (1 + short_lattice.loc[s,t-1]/100)
                else:
                    el_price.loc[s,t] = 0.5*((el_price.loc[s-1,t-1]/(1+short_lattice.loc[s-1,t-1]/100))+(el_price.loc[s,t-1]/(1+short_lattice.loc[s,t-1]/100)))

        bdt_zcb = el_price.sum(axis=0)
        bdt_spot = 100 * (((1/bdt_zcb)**(1/bdt_zcb.index))-1)

        # 차의 제곱의 합이 최소가 되도록 하는 a를 구하자.
        return ((bdt_spot.iloc[1:] - mkt_spot)**2).sum()

    a_init = [ 5 for i in range(14)]
    b_ = [(0,20) for i in range(len(a_init))]

    result = minimize(obj_function, a_init, bounds = b_)
    return result.x

def swaption_pricing(a, b, e):

    short_lattice = pd.DataFrame(columns = [ i for i in range(14)], index = [ i for i in range(13, -1, -1) ])
    short_lattice.loc[0,0] = a[0]
    for t in range(1, len(short_lattice.columns)):
        for s in range(t+1):
            short_lattice.loc[s,t] = a[t]*np.exp(b*s)

    short_lattice = short_lattice/100
    fix = 0.1165

    swap_price = short_lattice.loc[9:,:9].copy()
    swap_price.iloc[:,-1] = (swap_price.iloc[:,-1]-fix) / (1+swap_price.iloc[:,-1])
    for t in swap_price.columns[:-1][::-1]:
        for s in range(t+1):
            swap_price.loc[s,t] = 1/(1+short_lattice.loc[s,t])*((short_lattice.loc[s,t]-fix) + 0.5*(swap_price.loc[s+1,t+1] + swap_price.loc[s, t+1]))

    swaption_price = swap_price.copy()
    swaption_price[swaption_price<0] = 0
    swaption_price = swaption_price.loc[e:,:e]
    for t in swaption_price.columns[:-1][::-1]:
        for s in range(t+1):
            swaption_price.loc[s,t] = (swaption_price.loc[s+1,t+1] + swaption_price.loc[s,t+1])*0.5/(1+short_lattice.loc[s,t])
    
    return swaption_price.loc[0,0]*1000000

In [14]:
# 100만불 기준
mkt_spot = [7.3, 7.62, 8.1, 8.45, 9.2, 9.64, 10.12, 10.45, 10.75, 11.22, 11.55, 11.92, 12.2, 12.32]
b_list = [0.001, 0.005, 0.01, 0.05, 0.1]
e=2

for b in b_list:
    a = find_a(mkt_spot, b)
    price = swaption_pricing(a, b, e)
    print(f"b={b}일 때, Swaption Price는 ${price:.2f}이다.")

b=0.001일 때, Swaption Price는 $953.94이다.
b=0.005일 때, Swaption Price는 $1338.54이다.
b=0.01일 때, Swaption Price는 $1959.02이다.
b=0.05일 때, Swaption Price는 $6930.24이다.
b=0.1일 때, Swaption Price는 $13098.23이다.


---

## Problem 2.

Price the payer-swaption in the Coursera Lecture but now assume that it may be exercised at any time, $t ∈ {2,3, ⋯ ,9}$, and that the fixed rate in the underlying swap contract is now set at 11.65%.  
If exercised at time $t$ then the first cash flow occurs at $t + 1$ based on the short rate prevailing at time $t$. (Such an instrument is called a Bermudan swaption.)

In [15]:
mkt_spot = [7.3, 7.62, 8.1, 8.45, 9.2, 9.64, 10.12, 10.45, 10.75, 11.22, 11.55, 11.92, 12.2, 12.32]
b = 0.005
a = find_a(mkt_spot, b)
bermuda = pd.DataFrame()

for e in range(2,10):
    price = swaption_pricing(a, b, e)
    bermuda.loc['price', f't={e}'] = int(price)

In [16]:
bermuda

t=2      t=3      t=4      t=5      t=6      t=7      t=8      t=9
price  1338.0  21406.0  36900.0  33021.0  31773.0  24673.0  19536.0  13432.0

---

## Problem 3.

Construct a short lattice for periods (years) 0 through 9 with an initial rate of 6% and with successive rates determined by a multiplicative factor of either $u = 1.2$ or $d = 0.9$. Assign the risk-neutral probabilities to be 0.5.    

In [17]:
import pandas as pd
import numpy as np

u = 1.2
d = 0.9
q = 0.5
initial_rate = 0.06 
period = 10

In [18]:
df = pd.DataFrame(np.diag([initial_rate * d**i for i in range(period)]))
for i in range(period):
    for j in range(period):
        if j > i:
            df.iloc[i,j] = df.iloc[i,j-1]*1.2
df

0      1       2        3         4         5         6         7  \
0  0.06  0.072  0.0864  0.10368  0.124416  0.149299  0.179159  0.214991   
1  0.00  0.054  0.0648  0.07776  0.093312  0.111974  0.134369  0.161243   
2  0.00  0.000  0.0486  0.05832  0.069984  0.083981  0.100777  0.120932   
3  0.00  0.000  0.0000  0.04374  0.052488  0.062986  0.075583  0.090699   
4  0.00  0.000  0.0000  0.00000  0.039366  0.047239  0.056687  0.068024   
5  0.00  0.000  0.0000  0.00000  0.000000  0.035429  0.042515  0.051018   
6  0.00  0.000  0.0000  0.00000  0.000000  0.000000  0.031886  0.038264   
7  0.00  0.000  0.0000  0.00000  0.000000  0.000000  0.000000  0.028698   
8  0.00  0.000  0.0000  0.00000  0.000000  0.000000  0.000000  0.000000   
9  0.00  0.000  0.0000  0.00000  0.000000  0.000000  0.000000  0.000000   

          8         9  
0  0.257989  0.309587  
1  0.193492  0.232190  
2  0.145119  0.174143  
3  0.108839  0.130607  
4  0.081629  0.097955  
5  0.061222  0.073466  
6  0.045917  0.055100  
7  0.034437  0.041325  
8  0.025828  0.030994  
9  0.000000  0.023245

**(a) Using this lattice, find the value of a 10-year 6% bond.**

In [19]:
face_value = 100
c = face_value*0.06

df_pv = pd.DataFrame().reindex_like(df)
df_pv.iloc[:,-1] = (face_value + c) / (1+df.iloc[:,-1])

for i in range(period-2, -1, -1):
    for j in range(period):
        if i >= j:
            p_u = df_pv.iloc[j,i+1]
            p_d = df_pv.iloc[j+1,i+1]
            pv = (q*p_u + (1-q)*p_d)
            df_pv.iloc[j,i] = (pv + c) / (1+df.iloc[j,i])
df_pv

0          1           2           3           4           5  \
0  91.719355  84.975690   78.947553   73.773375   69.632201   66.774990   
1        NaN  97.469344   91.240326   85.763869   81.212196   77.816132   
2        NaN        NaN  102.225051   96.541528   91.653539   87.764405   
3        NaN        NaN         NaN  105.844848  100.690122   96.371235   
4        NaN        NaN         NaN         NaN  108.258881  103.579055   
5        NaN        NaN         NaN         NaN         NaN  109.462146   
6        NaN        NaN         NaN         NaN         NaN         NaN   
7        NaN        NaN         NaN         NaN         NaN         NaN   
8        NaN        NaN         NaN         NaN         NaN         NaN   
9        NaN        NaN         NaN         NaN         NaN         NaN   

            6           7           8           9  
0   65.582527   66.675509   71.132277   80.941560  
1   75.906357   75.988951   78.887989   86.025686  
2   85.152736   84.222727   85.595308   90.278643  
3   93.117125   91.245612   91.220652   93.754953  
4   99.765345   97.064729   95.822393   96.543101  
5  105.178748  101.776765   99.512614   98.745522  
6  109.501900  105.524139  102.425878  100.464430  
7         NaN  108.462917  104.697899  101.793403  
8         NaN         NaN  106.453232  102.813437  
9         NaN         NaN         NaN  103.591981

쿠폰 금리 6%, 만기 10년 채권의 현재가치는 약 **91.7194$**이다.

**(b) Suppose this bond can be called by the issuing party at any time after 5 years. (When the bond is called, the face value plus the currently due coupon are paid at that time and the bond is canceled.) What is the fair value of this bond?**

In [20]:
df_pv_call = pd.DataFrame().reindex_like(df)
df_pv_call.iloc[:,-1] = (face_value + c) / (1+df.iloc[:,-1])
df_call = df.copy()

for i in range(period-2, -1, -1):
    for j in range(period):
        if i >= j:
            if (df_call.iloc[j,i] < 0.06) & (i > 4):
                df_call.iloc[j,i] = 0.06
            p_u = df_pv_call.iloc[j,i+1]
            p_d = df_pv_call.iloc[j+1,i+1]
            pv = (q*p_u + (1-q)*p_d)
            df_pv_call.iloc[j,i] = (pv + c) / (1+df_call.iloc[j,i])
df_pv_call

0          1          2           3           4           5  \
0  90.956018  84.698301  78.899618   73.773375   69.632201   66.774990   
1        NaN  96.128457  90.693539   85.659716   81.212196   77.816132   
2        NaN        NaN  99.945249   95.481245   91.429035   87.764405   
3        NaN        NaN        NaN  102.123931   98.670387   95.890803   
4        NaN        NaN        NaN         NaN  102.511277   99.807993   
5        NaN        NaN        NaN         NaN         NaN  101.285478   
6        NaN        NaN        NaN         NaN         NaN         NaN   
7        NaN        NaN        NaN         NaN         NaN         NaN   
8        NaN        NaN        NaN         NaN         NaN         NaN   
9        NaN        NaN        NaN         NaN         NaN         NaN   

            6           7           8           9  
0   65.582527   66.675509   71.132277   80.941560  
1   75.906357   75.988951   78.887989   86.025686  
2   85.152736   84.222727   85.595308   90.278643  
3   93.117125   91.245612   91.220652   93.754953  
4   98.743960   97.064729   95.822393   96.543101  
5  100.848986  100.272467   99.512614   98.745522  
6  101.876227  101.527384  101.065015  100.464430  
7         NaN  102.450218  102.173038  101.793403  
8         NaN         NaN  103.021424  102.813437  
9         NaN         NaN         NaN  103.591981

단기 이자율이 쿠폰 이자율인 6%보다 낮으면 발행자는 채권을 회수하는 것이 유리하다.  

따라서 5년 후 단기 이자율이 6% 이하인 경우를 감안하여 계산하면 현재가치는 약 **90.9560$**이다. 

**(c) Use the forward equation to find the spot rate curve for the lattice.**  

In [21]:
df_forward = pd.DataFrame().reindex_like(df)
df_forward.iloc[0,0] = 1

for s in range(period):
    for k in range(s, period):
        if k == 0:
            continue
        elif s == 0:
            df_forward.iloc[s,k] = (q*df_forward.iloc[s, k-1]) / (1+df.iloc[s,k-1])
        elif s == k:
            df_forward.iloc[s,k] = (q*df_forward.iloc[s-1, k-1]) / (1+df.iloc[s-1,k-1])
        else:
            df_forward.iloc[s,k] = q*((df_forward.iloc[s-1, k-1])/ (1+df.iloc[s-1,k-1]) + df_forward.iloc[s,k-1]/(1+df.iloc[s,k-1]))
df_forward

0         1         2         3         4         5         6         7  \
0  1.0  0.471698  0.220008  0.101256  0.045872  0.020398  0.008874  0.003763   
1  NaN  0.471698  0.443774  0.309640  0.189521  0.107071  0.057019  0.028895   
2  NaN       NaN  0.223766  0.315081  0.292509  0.223361  0.151173  0.093799   
3  NaN       NaN       NaN  0.106697  0.199972  0.231688  0.212008  0.167222   
4  NaN       NaN       NaN       NaN  0.051113  0.119588  0.166077  0.177139   
5  NaN       NaN       NaN       NaN       NaN  0.024589  0.068971  0.111663   
6  NaN       NaN       NaN       NaN       NaN       NaN  0.011874  0.038832   
7  NaN       NaN       NaN       NaN       NaN       NaN       NaN  0.005753   
8  NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
9  NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

          8         9  
0  0.001549  0.000615  
1  0.013990  0.006476  
2  0.054281  0.029562  
3  0.118498  0.077134  
4  0.159586  0.127204  
5  0.136049  0.137872  
6  0.071822  0.098435  
7  0.021497  0.044725  
8  0.002796  0.011754  
9       NaN  0.001363

In [22]:
bond_price = df_forward.sum(axis=0)
bond_price

0    1.000000
1    0.943396
2    0.887548
3    0.832674
4    0.778987
5    0.726696
6    0.675995
7    0.627065
8    0.580068
9    0.535141
dtype: float64

In [23]:
spot_rates = (1/bond_price)**(1/bond_price.index) - 1
spot_rates

0    0.000000
1    0.060000
2    0.061461
3    0.062939
4    0.064431
5    0.065932
6    0.067438
7    0.068945
8    0.070447
9    0.071939
dtype: float64

## Problem 4.

What is the expected number of cards that need to be turned over in a regular 52-card deck in order to see the first ace?


에이스를 제외한 카드를 1~48이라 할 때,

$X_i =
\begin{cases}
1,\;\text {if card i is turned over before 4 aces}\\
0,\;\text {otherwise}
\end{cases}$

첫번째 A가 나올 때까지 뽑아야하는 카드 수 $X = 1 + \displaystyle \sum_{i=1}^{48}X$이고, 기댓값 $E[X] = 1 + \displaystyle \sum_{i=1}^{48}E[X_i]$ 이다.  

A 4장을 기준으로 카드 전체를 5구간으로 나누면 A 이전 구간에서 카드를 뽑을 확률은  $\displaystyle \frac {1}{5}$ 이다.

따라서 $E[X] = 1 + \displaystyle \sum_{i=1}^{48}E[X_i] = 1 + \frac {48} {5} = 10.6$

---

## Problem 5.

Solve the corresponding leetcode problem below and register the solution on GitHub.

**two sum**  

Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.  
You may assume that each input would have exactly one solution, and you may not use the same element twice.  
You can return the answer in any order.

In [24]:
nums = [2,7,11,15]
target = 9

In [25]:
class Solution5:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        for i in range(len(nums)):
            for j in range(i+1,len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]

In [26]:
Solution5().twoSum(nums, target)

[0, 1]

---

## Problem 6.

Solve the corresponding leetcode problem below and register the solution on GitHub.

**Valid Palindrome**

A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.

In [27]:
s = "A man, a plan, a canal: Panama"

In [28]:
class Solution6:
    def isPalindrome(self, s: str) -> bool:
        result = ''
        for c in s:
            if not (c.isalpha() or c.isnumeric()):
                continue
            elif c.isupper():
                result += c.lower()
            else:
                result += c
        return result == result[::-1]

In [29]:
Solution6().isPalindrome(s)

True

---